src/components/qTip/theme.css

```
@@ -23,11 +23,12 @@
   border-right: 0;
   border-top-left-radius: 4px;
   border-bottom-left-radius: 4px;
+  transform: translateX(100%) translateX(calc(-1 * var(--tip-width) + 1px)); /* split into 2 separete translate, otherwise IE can't handle it */
```

PP: translate3d is slightly more performant:
```suggestion
  transform: translate3d(100%, 0, 0) translate3d(calc(-1 * var(--tip-width) + 1px), 0, 0); /* split into 2 separete translate, otherwise IE can't handle it */
```

---

If we're worried about performance here (which I'm not convinced we need to here). 

I would prefer the will-change property over the 3d hack. 
It makes the actually translate easier to read and achieves the same result.

https://developer.mozilla.org/en-US/docs/Web/CSS/will-change

But again, as in the MDN docs.

> Important: will-change is intended to be used as a last resort, in order to try to deal with existing performance problems. It should not be used to anticipate performance problems.

---

I think it's weird that they say `It should not be used to anticipate performance problems.` and don't give the possibility to read up on why it shouldn't be used that way. To me it seems, anything that can anticipate performance problems should be implemented?

If I take an extreme comparison to explain why I feel that way:
> It's not a problem to do 3 for loops in each other here - once we notice there are performance problems, then we should tackle it 👍 	

Right? 😛

---

> To me it seems, anything that can anticipate performance problems should be implemented?

No cause that would cripple the speed at which you develop things, that would make PRs a constant battle of 'write it like this, cause this is more performant' which to me doesn't make sense and wastes a lot of time in a lot of cases.

> It's not a problem to do 3 for loops in each other here - once we notice there are performance problems, then we should tackle it 👍

I can't see anything wrong with 3 loops in each other when you're not anticipating a lot of data.

Everything depends on the context, as I meant with "should we be worried about performance here?". In the context of what we're doing now, it doesn't make sense to think about making a single small open and close animation ops/sec faster (cause that's what the translateX -> translate3d will achieve), while losing a bit of readability in the process).

Or in short, "premature optimization is the root of all evil".

---

But I see we merged it anyways 🤷‍♂️

---

I can agree with what you're saying, but I think that should not be a rule, more like a guideline... If it would be a big hassle for a small performance optimization then I would also just leave it. Otherwise I honestly can't understand why you would vouch for something less performant 🤔 	.

---

> But I see we merged it anyways 🤷‍♂️
Yeah idd 🤔 	...

---

So you think using something like `translate3d` and explaining other devs later in the project that it's not actually 3d but a trick to make it faster for no performance gain in this context is worth it?

---

I do in fact, if I got that explanation I would react with: "Oh okay, good to know 👍 	" 🤷‍♂️ 	And definitely worth it, because hopefully you only have to tell this person once...  Even in the smallest of projects I would use translate3d because you might as well just make a good habit out of using the most performant possibility there is, no?

---

I really strongly disagree with that statement because of the reasons above. 

I really don't get it :)

---

As always, if there's a better idea about something / a smarter or better approach for something, I'm always open to change my mind, but, just as you aren't, I'm not convinced yet 😬

---

Don't (prematurely) optimize for things that are not applicable or not perceived by the user. That's not an opinion, that's just common sense.

I think I outlined 'the smart or better approach' a couple of times in the previous comments. Optimize for **real cases**, **consider the context** and **keep the readability in mind**.

I don't have anything more to add to this discussion.

---

Also, please don't take my word on this topic. There are at least 1000 blog posts on premature optimization and why it's harmful. 

including this one. 
https://cdb.reacttraining.com/react-inline-functions-and-performance-bdff784f5578